aptos-move/aptos-vm/src/data_cache.rs

```diff
@@ -10,16 +11,98 @@ use aptos_types::{
     access_path::AccessPath,
     on_chain_config::ConfigStorage,
     state_store::{state_key::StateKey, state_storage_usage::StateStorageUsage},
-    vm_status::StatusCode,
 };
-use move_binary_format::errors::*;
+use move_binary_format::{errors::*, CompiledModule};
 use move_core_types::{
     account_address::AccountAddress,
     language_storage::{ModuleId, StructTag},
     resolver::{ModuleResolver, ResourceResolver},
+    vm_status::StatusCode,
 };
 use move_table_extension::{TableHandle, TableResolver};
-use std::ops::{Deref, DerefMut};
+use move_vm_runtime::move_vm::MoveVM;
+use std::{
+    collections::BTreeMap,
+    ops::{Deref, DerefMut},
+};
+
+pub struct MoveResolverWithVMMetadata<'a, 'm, S> {
+    move_resolver: &'a S,
+    move_vm: &'m MoveVM,
+}
+
+impl<'a, 'm, S: MoveResolverExt> MoveResolverWithVMMetadata<'a, 'm, S> {
+    pub fn new(move_resolver: &'a S, move_vm: &'m MoveVM) -> Self {
+        Self {
+            move_resolver,
+            move_vm,
+        }
+    }
+}
+
+impl<'a, 'm, S: MoveResolverExt> MoveResolverExt for MoveResolverWithVMMetadata<'a, 'm, S> {
+    fn get_resource_from_group(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+        resource_group: &StructTag,
+    ) -> Result<Option<Vec<u8>>, VMError> {
+        self.move_resolver
+            .get_resource_from_group(address, struct_tag, resource_group)
+    }
+
+    fn get_resource_group(&self, struct_tag: &StructTag) -> Result<Option<StructTag>, VMError> {
+        let metadata = aptos_framework::get_vm_metadata(self.move_vm, struct_tag.module_id());
+        Ok(Self::get_resource_group_from_metadata(struct_tag, metadata))
+    }
+
+    fn get_resource_group_data(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+    ) -> Result<Option<Vec<u8>>, VMError> {
+        self.move_resolver
+            .get_resource_group_data(address, struct_tag)
+    }
+}
+
+impl<'a, 'm, S: MoveResolverExt> ModuleResolver for MoveResolverWithVMMetadata<'a, 'm, S> {
+    type Error = VMError;
+
+    fn get_module(&self, module_id: &ModuleId) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.move_resolver.get_module(module_id)
+    }
+}
+
+impl<'a, 'm, S: MoveResolverExt> ResourceResolver for MoveResolverWithVMMetadata<'a, 'm, S> {
+    type Error = VMError;
+
+    fn get_resource(
+        &self,
+        address: &AccountAddress,
+        struct_tag: &StructTag,
+    ) -> Result<Option<Vec<u8>>, Self::Error> {
+        self.move_resolver.get_resource(address, struct_tag)
```

Comment:
This calls the underlying resolver's `get_resource()` so the cache is not really used?